go/database/mpt/proof_test.go

```diff
@@ -0,0 +1,341 @@
+							storage: &Branch{
+								children: Children{
+									key1Path[0]: &Extension{path: key1Path[1:45],
+										next: &Value{key: key1, length: 19, value: common.Value{0x12}}},
+									key2Path[0]: &Extension{path: key2Path[1:45],
+										next: &Value{key: key2, length: 19, value: common.Value{0x34}}},
+								},
+							},
+						}},
+					}},
+				},
+			}},
+		}},
+	})
+
+	rootHandle := node.GetReadHandle()
+	rootHash, dirty := rootHandle.Get().GetHash()
+	if dirty {
+		t.Fatalf("expected node to be clean")
+	}
+	rootHandle.Release()
+
+	// create following shadow proofs
+	// 1. proof that contains only nodes for address1 and key1
+	// 2. proof that contains only nodes for address2
+	// 3. wide proof that includes both previous proofs
+	// 4. total proof that includes all nodes in the tree
+	address1Key1ProofShadow := testCreateShadowProofForLabels(t, ctxt, "R", "A_1", "A_2", "A_3", "K_1", "K_2", "K_3")
+	address2ProofShadow := testCreateShadowProofForLabels(t, ctxt, "R", "B_1", "B_2", "B_3")
+	wideProof := testCreateShadowProofForLabels(t, ctxt, "R", "A_1", "A_2", "A_3", "K_1", "K_2", "K_3", "B_1", "B_2", "B_3")
+	totalProof := testCreateShadowProof(t, ctxt, &root, node)
+
+	// make sure extracted proofs are different
+	if address1Key1ProofShadow.Equals(address2ProofShadow) {
+		t.Fatalf("address1 and address1Key1 proofs are equal")
+	}
+	if wideProof.Equals(address2ProofShadow) {
+		t.Fatalf("total and address1 proofs are equal")
+	}
+	if wideProof.Equals(address1Key1ProofShadow) {
+		t.Fatalf("total and address1Key1ProofShadow proofs are equal")
+	}
+	if wideProof.Equals(totalProof) {
+		t.Fatalf("total and wide proofs are equal")
+	}
+
+	t.Run("Extract", func(t *testing.T) {
+		// Test proofs can be extracted and that they match the shadow proofs
+		extractedProofAddress1Key1, exists := totalProof.Extract(rootHash, address1, key1)
+		if !exists {
+			t.Errorf("proof for %v %v %v not found", rootHash, address1, key1)
+		}
+		if got, want := extractedProofAddress1Key1, address1Key1ProofShadow; !got.Equals(want) {
+			t.Errorf("unexpected proof: got %v, want %v", got, want)
+		}
+
+		extractedProofAddress2, exists := totalProof.Extract(rootHash, address2)
+		if !exists {
+			t.Errorf("proof for %v %v %v not found", rootHash, address1, key1)
+		}
+		if got, want := extractedProofAddress2, address2ProofShadow; !got.Equals(want) {
+			t.Errorf("unexpected proof: got %v, want %v", got, want)
+		}
+	})
+
+	t.Run("Merge", func(t *testing.T) {
+		mergedProof := MergeProofs(address1Key1ProofShadow, address2ProofShadow)
+		if got, want := mergedProof, wideProof; !got.Equals(want) {
+			t.Errorf("unexpected proof: got %v, want %v", got, want)
+		}
+	})
+}
+
+func TestWitnessProof_Extract_Various_NodeTypes_NotFoundProofs(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	address := common.Address{0xAB, 0xCD, 0xEF}
+	key := common.Key{0x12, 0x34, 0x56, 0x78}
```

Comment:
Same here, the test basically works with any address or key.
```suggestion
	address := common.Address{}
	key := common.Key{}
```